File: iq/list/singly/midpoint/find.go
// Copyright 2025 Samvel Khalatyan. All rights reserved.

package midpoint

type Node struct {
	Val  int
	Next *Node
}

func Find(node *Node) *Node {
	it := node
	for node != nil && node.Next != nil {
		node = node.Next.Next
		it = it.Next
	}
	return it
}

File: iq/list/singly/midpoint/find.cc
// Copyright 2025 Samvel Khalatyan. All rights reserved.
//
// clang-format off-next-line
//go:build ignore

#include "iq/list/singly/midpoint/find.h"

#include <memory>

namespace iq::list::singly::midpoint {

std::shared_ptr<Node> Find(std::shared_ptr<Node> list) {
  std::shared_ptr<Node> slow = list;
  for (std::shared_ptr<Node> fast = list; fast && fast->next;
       fast = fast->next->next) {
    slow = slow->next;
  }
  return slow;
}

}  // namespace iq::list::singly::midpoint
